packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1164,32 +1213,179 @@ export function errorHydratingContainer(parentContainer: Container): void {
   }
 }
 
-export function getInstanceFromNode(node: HTMLElement): null | Object {
-  return getClosestInstanceFromNode(node) || null;
+export function isPersistentInstance(instance: Instance | Container): boolean {
+  if (instance.nodeType === ELEMENT_NODE) {
+    return isPersistentInstanceType(instance.tagName.toLowerCase());
+  }
+  return false;
 }
 
-export function preparePortalMount(portalInstance: Instance): void {
-  listenToAllSupportedEvents(portalInstance);
+function isPersistentInstanceType(type: string): boolean {
+  switch (type) {
+    case HTML:
+    case HEAD:
+    case BODY: {
+      return true;
+    }
+    default: {
+      return false;
+    }
+  }
 }
 
-export function prepareScopeUpdate(
-  scopeInstance: ReactScopeInstance,
+export function createInsertionParentInstance(
+  fromInstance: Instance,
+): InsertionParentInstance {
+  return fromInstance.ownerDocument.createDocumentFragment();
+}
+
+export function commitPersistentPlacement(
+  type: string,
+  props: Props,
+  insertionFragment: DocumentFragment,
+  fallbackInstance: Instance,
+  rootContainerInstance: Container,
   internalInstanceHandle: Object,
-): void {
-  if (enableScopeAPI) {
-    precacheFiberNode(internalInstanceHandle, scopeInstance);
+): Element {
+  const ownerDocument = insertionFragment.ownerDocument;
+  // For the three persistent Host Components that exist in DOM it is necessary for there to
+  // always be a documentElement. This should always be the case but for the sake of correctness
+  // we check and construct one if one does not yet exist.
+  // Additionally we assume HTML document here because the only persistent host elements are when using
+  // this kind of document
+  if (!ownerDocument.documentElement) {
+    ownerDocument.append(ownerDocument.createElement(HTML));
+  }
+  const documentElement: HTMLHtmlElement = (ownerDocument.documentElement: any);
+  switch (type) {
+    case HTML: {
+      resetProperties(documentElement, type, props, rootContainerInstance);
+      precacheFiberNode(internalInstanceHandle, documentElement);
+      updateFiberProps(documentElement, props);
+      // @TODO figure out how to make this support innerHTML
+      if (__DEV__) {
+        if (insertionFragment.children.length > 0) {
+          throw new Error(
+            `The <${HTML}> element had insertion children when none were expected. This is a bug in React.`,
+          );
+        }
+      }
+      return documentElement;
+    }
+    case HEAD:
+    case BODY: {
+      // $FlowIssue Flow doens't know that body and head are the properties being accessed here
+      let element = ownerDocument[type];
+      if (element) {
+        resetPersistentInstance(element);
+        resetProperties(element, type, props, rootContainerInstance);
+        precacheFiberNode(internalInstanceHandle, element);
+        updateFiberProps(element, props);
+      } else {
+        element = fallbackInstance;
```

Comment:
At the moment the fallback instance already has props set on it but that is technically wasted work almost all the time. we could defer it to here but the HostContext isn't built up in the commit phase so it's potentially harder to do here